File: lt0665.py
# https://leetcode-cn.com/problems/non-decreasing-array/ 
# non decreasing array 
class Solution:
    def checkPossibility(self, nums: List[int]) -> bool:

        n = len(nums)

        if n == 1:
            return True

        incorrect = 0

        if nums[0] > nums[1]:
            nums[0] = nums[1]
            incorrect += 1
        
        for i in range(1, n - 1):

            r = nums[i + 1]
            if nums[i] > r:
                incorrect += 1
                if incorrect > 1:
                    return False
                l = nums[i - 1]

                if l <= r:
                    nums[i] = l
                else:
                    nums[i + 1] = nums[i]
                
        return True
                

File: code/lt0665.java
class Solution {
    public boolean checkPossibility(int[] nums) {
        int len = nums.length;
        if(len < 3)
            return true;
        int count = 0;
        if(nums[0] > nums[1]){
            nums[0] = nums[1];
            count++;
        }
        
        for(int i = 1; i < len - 1; i++){
            int right = nums[i + 1];
            if(nums[i] > right){
                count++;
                if(count > 1)
                    return false;
                
                int left = nums[i - 1];
                if(left > right)
                    nums[i+1] = nums[i];
                else
                    nums[i] = left;
            }
        }
        
        return true;
    }
}